tests/test_density_matrix.py

```diff
@@ -0,0 +1,397 @@
+
+        assert np.allclose(backend.state.rho, expected_matrix_1) or np.allclose(backend.state.rho, expected_matrix_2)
+
+    def test_correct_byproduct(self):
+        circ = Circuit(1)
+        circ.rx(0, np.pi / 2)
+        pattern = circ.transpile()
+
+        backend = DensityMatrixBackend(pattern)
+        backend.add_nodes([0, 1, 2])
+        backend.entangle_nodes((0, 1))
+        backend.entangle_nodes((1, 2))
+        backend.measure(backend.pattern.seq[-4])
+        backend.measure(backend.pattern.seq[-3])
+        backend.correct_byproduct(backend.pattern.seq[-2])
+        backend.correct_byproduct(backend.pattern.seq[-1])
+        backend.finalize()
+        rho = backend.state.rho
+
+        backend = StatevectorBackend(pattern)
+        backend.add_nodes([0, 1, 2])
+        backend.entangle_nodes((0, 1))
+        backend.entangle_nodes((1, 2))
+        backend.measure(backend.pattern.seq[-4])
+        backend.measure(backend.pattern.seq[-3])
+        backend.correct_byproduct(backend.pattern.seq[-2])
+        backend.correct_byproduct(backend.pattern.seq[-1])
+        backend.finalize()
+        psi = backend.state.psi
+
+        np.testing.assert_allclose(rho, np.outer(psi, psi.conj()))
+
+    def test_dephase(self):
+        def simulate(p, pattern):
+            backend = DensityMatrixBackend(pattern)
+            backend.add_nodes([0])
+            backend.dephase(p)
+            backend.add_nodes([1])
+            backend.dephase(p)
+            backend.add_nodes([2])
+            backend.dephase(p)
+            backend.entangle_nodes((0, 1))
+            backend.dephase(p)
+            backend.entangle_nodes((1, 2))
+            backend.dephase(p)
+            backend.measure(pattern.seq[-4])
+            backend.dephase(p)
+            backend.measure(pattern.seq[-3])
+            backend.dephase(p)
+            backend.correct_byproduct(pattern.seq[-2])
+            backend.dephase(p)
+            backend.correct_byproduct(pattern.seq[-1])
+            backend.dephase(p)
+            backend.finalize()
+            return backend
+
+        circ = Circuit(1)
+        circ.rx(0, np.pi / 4)
+        pattern = circ.transpile()
+        backend1 = simulate(0, pattern)
+        backend2 = simulate(1, pattern)
+        np.testing.assert_allclose(backend1.state.rho, backend2.state.rho)
+
+        circ = Circuit(1)
+        circ.rz(0, np.pi / 3)
+        pattern = circ.transpile()
+        dm_backend = simulate(1, pattern)
+        sv_backend = StatevectorBackend(pattern)
+        sv_backend.add_nodes([0, 1, 2])
+        sv_backend.entangle_nodes((0, 1))
+        sv_backend.entangle_nodes((1, 2))
+        sv_backend.measure(pattern.seq[-4])
+        sv_backend.measure(pattern.seq[-3])
+        sv_backend.correct_byproduct(pattern.seq[-2])
+        sv_backend.correct_byproduct(pattern.seq[-1])
+        sv_backend.finalize()
+        np.testing.assert_allclose(dm_backend.state.fidelity(sv_backend.state.psi), 0.25)
+
+
```

Comment:
could you add tests for larger simulations (see tests for pattern and sv sim)? compare results with circuit.simulat_statevector at no-noise case, and with noisy case, fix the np.random.seed and check the results with some known result you calculate without using the backend?

Reply:
I added the test case for 3-qubit QFT. Please see that if it's enough.

Reply:
could you use https://github.com/TeamGraphix/graphix/blob/master/tests/random_circuit.py#L56 instead?

Reply:
After the commit [a6745bb](https://github.com/TeamGraphix/graphix/pull/64/commits/a6745bb6e09a039bab0d8b30d88f7d79294d7165), I'll fix this part.

Reply:
What do you mean by this? I think it's not the test for `dephase` if you mean something like `np.testing.assert_allclose(dm_backend.state.fidelity(random_state), dm_backend.state.fidelity(exact_qft_state))` instead of
`np.testing.assert_allclose(dm_backend.state.fidelity(noisy_state), dm_backend.state.fidelity(exact_qft_state))`

Reply:
I mean something like `np.testing.assert_allclose(1, dm_backend.state.fidelity(known_noisy_state_for_noise_realization))` (fix the noise realizatino by np.random.seed())